### ISSUE TITLE
Question: role of MSSpectrum in MRMTransitionGroup

### ISSUE THREAD

**Issue by timosachsenberg** [OpenMS/OpenMS#7725]
What is the motivation of having an MSSpectrum as template parameter in: 

MRMTransitionGroup<**MSSpectrum**, ReactionMonitoringTransition> trgroup;
used in TransitionGroupOpenMS

**Comment by timosachsenberg** [comment]
In general, I wonder why TransitionGroupOpenMS and also other classes have been introduced.
Do you guys know?
One reason (just guess) could be that Hannes wanted to use slimmer datastructures at some point and introduced
an interface to ease the transition from the bigger (e.g. more generic, OpenMS ones) to slimmer ones.

**Comment by jcharkow** [comment]
> What is the motivation of having an MSSpectrum as template parameter in:
> 
> MRMTransitionGroup<**MSSpectrum**, ReactionMonitoringTransition> trgroup; used in TransitionGroupOpenMS

I am not entirely sure but one possible reason was because some of the scores (e.g. DIA scores) require the spectrum associated with the transition and not just the chromatogram.

**Comment by jcharkow** [comment]
> In general, I wonder why TransitionGroupOpenMS and also other classes have been introduced. Do you guys know? One reason (just guess) could be that Hannes wanted to use slimmer datastructures at some point and introduced an interface to ease the transition from the bigger (e.g. more generic, OpenMS ones) to slimmer ones.

Yes I think that slimming them down has to do with things. The other reason is that OpenSWATH has its own interface for dealing raw data which supports various file formats including .cached_mzml, .mzml, .sqMass etc. This likely came about because unlike most OpenMS algorithms, it is usually unfeasible to store the entire DIA file in memory so there needed to be interfaces to access data no matter how it is stored.

**Comment by jcharkow** [comment]
We were talking about this a while ago with @poshul of extending this abstracted file access support across the MSExperiment object.

E.g.  MSExperiment would be a parent class to OnDiscMSExperiment, CachedMSExperiment, InMemoryMSExperiment or something like that. 

There was some reasons why we did not adopt this that I cannot remember. Possibly it was just too much work to do without enough benefit.

**Comment by timosachsenberg** [comment]
I did a quick search in my editor and if I did not overlook something.

- doesn't seem to be used:

TransitionGroupOpenMS

- seem to be used:

FeatureOpenMS
SignalToNoiseOpenMS
MRMFeatureOpenMS

but mainly to be able to introduce MockObjects.
e.g. search for 
public OpenSwath::IMRMFeature
public OpenSwath::IFeature
public OpenSwath::ITransitionGroup

- except for:
OpenSwath::ISignalToNoise 
which has noch MockObjects but this is also only used to execute SignalToNoiseOpenMS through a pointer. 
Did not see any use of polymorphism here.

**Comment by timosachsenberg** [comment]
> We were talking about this a while ago with @poshul of extending this abstracted file access support across the MSExperiment object.
> 
> E.g. MSExperiment would be a parent class to OnDiscMSExperiment, CachedMSExperiment, InMemoryMSExperiment or something like that.
> 
> There was some reasons why we did not adopt this that I cannot remember. Possibly it was just too much work to do without enough benefit.

Yeah, it would entail a lot of rewrites. I think inheriting from some pure virtual Spectrum/Chromatogram/Metadata/AccessInterface in each of these classes would be feasible but still some work.

**Comment by jcharkow** [comment]
Hmm maybe these were just for testing purposes? I can ask Hannes later this week

**Comment by timosachsenberg** [comment]
But regarding the classes and interfaces I listed above... I don't think these are handled out of memory right now. 
So if we want to simplify some code, we could consider removing those (or agree that we will need them anyway in the future).
Yeah, probably best to ask Hannes. There might be good reasons to keep them.

**Comment by jcharkow** [comment]
Yes would be good to remove them if they are only for testing purposes, especially if we can make the datatypes simpler. I'll let you know later this week

**Comment by timosachsenberg** [comment]
thanks!
Just some ideas so they don't get forgotten:
e.g. instead of
```
  class OPENMS_DLLAPI FeatureOpenMS :
    public OpenSwath::IFeature
  {
public:

    explicit FeatureOpenMS(Feature& feature);

    ~FeatureOpenMS() override;

    void getRT(std::vector<double>& rt) const override;

    void getIntensity(std::vector<double>& intens) const override;

    float getIntensity() const override;

    double getRT() const override;

private:
    Feature* feature_;
  };
```
We could simplify by using Peak2D instead of IFeature, as Peak2D already provides retention time (RT), m/z, and intensity. 
For instance:
- Instead of introducing an intermediate class, we might be able to directly use a Peak2D* with a Feature. Because Feature derives from Peak2D
- If necessary, we can even perform type slicing to reduce the data size, for example:
Peak2D my_sliced_object = some_feature; 
creates a sliced copy of the Peak2D part of a feature.

**Comment by hroest** [comment]
maybe I can help answer some questions

- MRMTransitionGroup - this was introduced because there was no data structure to represent a collection of chromatograms as well as a relationship to their assays

-  MRMFeatureAccessOpenMS.h - this was created after some of the main work on the algorithm in an attempt to decouple algorithms from implementations, however this was never fully achieved. The idea was that the algorithm part would not need to know about the implementation of MSSpectrum/MSChromatogram/MSFeature in order to perform its task and compute cross-correlation matrices etc. In the end its a question of abstraction vs complexity, the current implementation has more levels of abstraction at the cost of added complexity. 

- removing them would be some work since most of them are used, for example `OpenSwath::IFeature` is used in https://github.com/OpenMS/OpenMS/blob/develop/src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMScoring.h#L62 and is then used in the algorithm implementation.
